kale/pipeline/mida_trainer.py

```diff
@@ -0,0 +1,200 @@
+# =============================================================================
+# Author: Shuo Zhou, shuo.zhou@sheffield.ac.uk
+# =============================================================================
+
+"""Implementation of MIDA->Feature Selection->Linear SVM/LogisticRegression Pipeline
+
+References:
+    [1] Yan, K., Kou, L. and Zhang, D., 2018. Learning domain-invariant subspace using domain features and
+        independence maximization. IEEE transactions on cybernetics, 48(1), pp.288-299.
+"""
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.model_selection import GridSearchCV, LeaveOneGroupOut, ParameterGrid
+from sklearn.preprocessing import FunctionTransformer
+
+from ..embed.factorization import MIDA
+
+
+class MIDATrainer(BaseEstimator, ClassifierMixin):
+    """Trainer of pipeline: Transformer-> MIDA -> Estimator
+
+    Args:
+        estimator (BaseEstimator): Estimator object implementing 'fit'
+        transformer (BaseEstimator, optional): Transformer object implementing 'fit_transform'. Defaults to None.
+        mida_param_grid (dict, optional): Grids for searching the optimal hyper-parameters of MIDA. Defaults to None.
+        transformer_param_grid (dict, optional): Grids for searching the optimal hyper-parameters of the transformer.
+            Defaults to None.
+        estimator_param_grid (dict, optional): Grids for searching the optimal hyper-parameters of the estimator.
+            Defaults to None.
+    """
+
+    def __init__(
+        self,
+        estimator,
+        estimator_param_grid,
+        transformer=None,
+        mida_param_grid=None,
+        transformer_param_grid=None,
+    ):
+        self.estimator = estimator
+        if transformer is None:
+            self.transformer = FunctionTransformer()
+        else:
+            self.transformer = transformer
+        self.best_mida = MIDA()
+        if mida_param_grid is None:
+            self.mida_param_grid = {key: [val] for key, val in self.best_mida.get_params().items()}
+        else:
+            self.mida_param_grid = mida_param_grid
+        if transformer_param_grid is None:
+            self.transformer_param_grid = {key: [val] for key, val in self.transformer.get_params().items()}
+        else:
+            self.transformer_param_grid = transformer_param_grid
+        if estimator_param_grid is None:
+            self.estimator_param_grid = {key: [val] for key, val in self.estimator.get_params().items()}
+        else:
+            self.estimator_param_grid = estimator_param_grid
+        self.best_mida_params = None
```

Comment:
Is using a dictionary to save and update these parameters better and more compact?  We can talk about it. 
```suggestion
        self.best_params = {
              'mida_params': None,
              'transformer': None,
              'transformer_params': None,
              'estimator': None,
              'estimator_params': None,
              'score': None
}
```